xhttp/client.go

```diff
@@ -0,0 +1,144 @@
+// On such a situation you can have two timeouts. One provided on the request passed to [Client.Do] on the request context and the timeout
+// defined with this option. Lets say the overall timeout is 10 min (when you created the original request) and this configuration here is
+// 30 secs. Now every 30 sec the request will fail since it timeouted and will be retried, until the parent timeout of 10 min expires.
+func RetrierWithRequestTimeout(timeout time.Duration) RetrierOption {
+	return func(r *retrierClient) {
+		r.requestTimeout = timeout
+	}
+}
+
+// NewRetrierClient wraps the given client with retry logic.
+// The returned [Client] will automatically retry failed requests.
+func NewRetrierClient(c Client, options ...RetrierOption) Client {
+	r := &retrierClient{
+		client:    c,
+		sleep:     defaultSleep,
+		minPeriod: time.Second,
+	}
+	for _, option := range options {
+		option(r)
+	}
+	return r
+}
+
+type retrierClient struct {
+	client         Client
+	requestTimeout time.Duration
+	minPeriod      time.Duration
+	sleep          func(context.Context, time.Duration)
+}
+
+func (r *retrierClient) Do(req *http.Request) (*http.Response, error) {
+	requestBody, err := io.ReadAll(req.Body)
+	if err != nil {
+		return nil, fmt.Errorf("reading request body: %w", err)
+	}
+	if err := req.Body.Close(); err != nil {
+		return nil, fmt.Errorf("closing request body: %w", err)
+	}
+	return r.do(req.Context(), req, requestBody, r.minPeriod)
+}
+
+func (r *retrierClient) do(ctx context.Context, req *http.Request, requestBody []byte, sleepPeriod time.Duration) (*http.Response, error) {
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+	req, cancel := r.newRequest(ctx, req, requestBody)
+	defer cancel()
+
+	res, err := r.client.Do(req)
+	if err != nil {
+		// Sadly there is no other way to detect this error other than using the opaque string message
+		// The error type is internal and the http pkg does not provide a way to check it
+		// - https://cs.opensource.google/go/go/+/refs/tags/go1.21.4:src/net/http/h2_bundle.go;l=9250
+		//
+		// For connections reset...same problem:
+		// - https://github.com/golang/go/blob/d0dc93c8e1a5be4e0a44b7f8ecb0cb1417de50ce/src/net/http/transport_test.go#L2207
+		// We need to go for some suffix matches.
+		if strings.Contains(err.Error(), "http2: server sent GOAWAY and closed the connection") ||
+			strings.HasSuffix(err.Error(), ": connection reset by peer") {
+
+			r.sleep(ctx, sleepPeriod)
+			return r.do(ctx, req, requestBody, sleepPeriod*2)
+		}
+		return nil, err
+	}
+
+	if res.StatusCode == http.StatusInternalServerError ||
+		res.StatusCode == http.StatusServiceUnavailable ||
+		res.StatusCode == http.StatusTooManyRequests {
+		// Maybe add handling for Retry-After header, so far this seems to be enough
+		r.sleep(ctx, sleepPeriod)
+		return r.do(ctx, req, requestBody, sleepPeriod*2)
```

Comment:
Will we retry forever? Also maybe the status could be customized, for instance in the API you're integrating Conflict should be retried as well.

Reply:
I was not sure if we can generalize Conflict to always retry...but if you think so we can add it. Regarding re-trying forever, no, it is controlled by the context added on the request (how you normally would create a timeout when making a request using http.Client). If no context is defined, then yeah it is forever (which is the same behavior of the http.Client since it can literally hang forever waiting for an answer).

We can add an extra max retries thing... so far I didn't use it, I just always define a context with a timeout and use it when sending requests (and that controls the worse case of how long I will wait for a response).

Reply:
In general I think about retrying more in terms of how long I'm willing to wait than how much times did I retry already, because there is no correlation with how much times you tried already with being safe to give up, so I just try until I have time to do it.

Reply:
I don't think we should add Conflict here, most cases we shouldn't retry. My point was about be configurable.

Regarding the ctx, I understand, but not sure if time will always be a good metric to stop. Nevertheless, if you don't need for your use-case we can always extend later.

Reply:
For Conflict Im not sure if in general just retrying with the exact same request always works... so I was less prone to define it here...maybe it could be an option ? (it could be an option that keeps adding status code that you also want to retry, and retries on any of them), like:

```go
xhttp.NewRetrierClient(c, xhttp.RetrierWithRetryOn(http.StatusConflict), xhttp.RetrierWithRetryOn(http.StatusSomethingElse))
```

Reply:
Yep! That was what I had in mind! 👍🏻 
Feel free to do later if someone needs it.

Reply:
> I don't think we should add Conflict here, most cases we shouldn't retry. My point was about be configurable.
> 
> Regarding the ctx, I understand, but not sure if time will always be a good metric to stop. Nevertheless, if you don't need for your use-case we can always extend later.

Ah yeah I missed that you already mentioned customization, yeah I think that makes sense and can be done :-).
Regarding the max retries, what is the scenario that you think it would be interesting ? we already have an exponential backoff, so even if we retry multiple times the sleep gets longer and longer, what would be the max retries that you would think is a good idea to give up ? Every time I'm integrating I always want to retry for as long as I can, until I have no more time, or else..what is a good number of retries ? 3 ? 10 ?

If a service is down for too long and we want to avoid sending requests to it I would say the solution is a circuit breaker, but I surivived so long just with exponential backoffs that Im not quite sure we need them.

Reply:
But Im not against adding it as a parameter, and by default we don't use it (and tries until context expires)

Reply:
I just feel that retrying by time will also include time used by the server, let's say that your request takes 2, 3 secs, how many retries fit there, don't really know, may fit a lot (which can be a problem as we're causing loading in a service that is already down), and may be too less. I feel that I retry X times more clear of my intentions comparing with I retry for X secs that again it may be too much or too little.

Reply:
For the problem of causing load to a service I believe a circuit breaker is the proper solution, having max retries only alleviates the problem (barely, if you retry a lot, and if you retry too little you risk failing unnecessarily in some cases).

Then regarding feeling better with retrying N times, I usually feel better retrying for N period of time =P. Retrying N times has the advantage that it may send less requests to a service that is down, but that is not a huge problem since we have an exponential backoff...we will not just hammer down like crazy a downed service (+ the proper solution is the circuit breaker if this is a huge problem). But it has the disadvantage of, lets say you configured 5 max retries...what if the sixth time it would work ? and the timeout you had configured to perform the job was 30 secs and you still had 29 seconds available ? why fail ? in this case you just failed the request unnecessarily, you had plenty of time to finish it and the server just got back up. When thinking in those terms, I never know how much retries will be the best...5 ? what if 6th  would work ? 10 ? what if the 11th would work ? and so goes on =P.

So by default I think no max retries and having a backoff protocol makes more sense in order to optimize to working instead of failing unnecessarily.

Reply:
Also, for a time based definition + exponential backoff  we can know the worse case scenario of how much requests will happen based on the sleep period. For a service that is down and fails immediately. If the context is configured to wait for 10 seconds, the worse case scenario is (for the default 1 second of min sleep period and considering the request fails immediately):

- first request
- 1 sec 
- second request
- 2 sec
- third request
- 4 sec
- fourth request
- 8 sec

It will never make more than 4 requests (because of the exponential backoff + min sleep period). We never have an unbounded number of requests and it is always possible to compute the worse case scenario. It is only unbounded if the context is unbounded (no time limit/cancellation), but then that is the caller option.

Reply:
Yep, it seems fine, then!